### PR TITLE
Add TP-plan for muse_glimmer and fix dflash for sharded embedding

### PR DESCRIPTION
**Pull request by sywangyi**
[CPU CI](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=47913) [GPU run-slow](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=47913)

…or parallelism and reduces memory pressure during DFlash assisted decoding.

Main changes:

- Add selective hidden-state capture via `output_hidden_states_layers`, so DFlash only stores the main-model hidden states needed by the assistant instead of retaining all layers.
- Update `DFlashTokenCandidateGenerator` to request only the configured target layers and prepare the assistant context from those sparse hidden states.
- Support TP-sharded DFlash assistant context when the assistant encoder projection uses `rowwise` tensor parallelism, avoiding materializing the full concatenated context on every rank.
- Fix DFlash with MuseGlimmer TP by letting its normed embedding module expose an `apply_norm=False` path, so DFlash can obtain raw noise-token embeddings without bypassing TP-aware embedding dispatch.
- Add default tensor-parallel plans for MuseGlimmer text, vision, multimodal projection, LM head, and MuseGlimmer assistant modules.

These changes allow DFlash to work correctly with MuseGlimmer/MuseGlimmer assistant in TP setups while reducing unnecessary hidden-state memory usage.

With these changes, MuseGlimmer DFlash generation works on 2 x 32GB GPUs for both text-only and image+text inputs.

 @ArthurZucker @CyrilVallez

### CODE REVIEW

**Comment by sywangyi** [comment]
@ameerhmz

**Comment by sywangyi** [comment]
Video inputs also benefit from lower memory pressure when multimodal embeddings are precomputed before generation.

File: src/transformers/models/muse_glimmer/configuration_muse_glimmer.py [hunk 1]
```diff
@@ -54,6 +54,15 @@ class MuseGlimmerVisionConfig(PreTrainedConfig):
     hidden_act: str = "gelu"
     rope_parameters: dict | None = None  # defaults set by `RopeConfigMixin`
     max_position_embeddings: int = 32 * 32  # == `pos_h * pos_w`
+    base_model_tp_plan = {
```
Comment (ArthurZucker) [inline src/transformers/models/muse_glimmer/configuration_muse_glimmer.py:57]: IDK if this is super super worth it, did you test it ?

File: src/transformers/generation/candidate_generator.py [hunk 2]
```diff
@@ -1597,6 +1601,41 @@ def __init__(
 
         self.is_main_model_prefill = True
 
+    def _prepare_noise_embeds(self, noise_ids: torch.LongTensor) -> torch.Tensor:
+        if hasattr(self.main_model_input_embeddings, "embed_norm"):
+            return self.main_model_input_embeddings(noise_ids, apply_norm=False)
+        return torch.nn.functional.embedding(noise_ids, self.main_model_input_embeddings.weight)
```
Comment (ArthurZucker) [inline src/transformers/generation/candidate_generator.py]: Why is there a need for if else here?

File: src/transformers/models/muse_glimmer/modeling_muse_glimmer.py [hunk 3]
```diff
@@ -440,8 +440,11 @@ def __init__(self, num_embeddings: int, embedding_dim: int, padding_idx: int, no
         # to embed without the norm
         self.embed_norm = MuseGlimmerRMSNorm(eps=norm_eps, with_scale=False)
 
-    def forward(self, input_ids: torch.Tensor):
-        return self.embed_norm(super().forward(input_ids))
+    def forward(self, input_ids: torch.Tensor, apply_norm: bool = True):
+        inputs_embeds = super().forward(input_ids)
+        if not apply_norm:
+            return inputs_embeds
+        return self.embed_norm(inputs_embeds)
```
Comment (ArthurZucker) [inline src/transformers/models/muse_glimmer/modeling_muse_glimmer.py]: this is a bit weird

File: src/transformers/utils/output_capturing.py
Comment (ArthurZucker) [inline src/transformers/utils/output_capturing.py:1]: can you explaiin the reason you need this?

File: src/transformers/generation/candidate_generator.py [hunk 4]
```diff
@@ -1597,6 +1601,41 @@ def __init__(
...
+
+    def _prepare_context_hidden_states(
```
Comment (ArthurZucker) [inline src/transformers/generation/candidate_generator.py]: we should not have assistance / decice specific code here. This usually means you need an extra module / tp plan mapping with new instructions. We never have TP code here. You should create a new plan / update the plan for the layer.

**Comment by Cyrilvallez** [comment]
A few thoughts here!
- we should really open differents PRs for differents things that we want to improve. For example, allowing capture of specific layer indices has been on the road map already, and should be done cleaner than this to be reusable nicely
- In general we don't want any TP-specific code. I don't really get why we need it here. If the TP plan is coherent, I believe we could reuse the outputs of the main model directly into the assistant

File: src/transformers/models/muse_glimmer/modeling_muse_glimmer.py (hunk 3 above)
Comment (sywangyi) [reply src/transformers/models/muse_glimmer/modeling_muse_glimmer.py]: DFlash needs raw token embeddings for the assistant noise tokens, not the normalized embeddings used by the main MuseGlimmer text path. apply_norm=False keeps the call going through the embedding module, which is important for TP-aware embedding lookup, while skipping only the main-model embedding norm.

File: src/transformers/utils/output_capturing.py
Comment (sywangyi) [reply src/transformers/utils/output_capturing.py:1]: The reason for this change is to avoid materializing and keeping all hidden states during DFlash generation. DFlash only needs the hidden states from assistant_model.config.target_layer_ids to build the assistant context, but output_hidden_states=True currently makes the main model retain every layer output. For MuseGlimmer this is expensive, especially in long-context / multimodal generation where video or image embeddings have already increased the sequence length. output_hidden_states_layers lets the generation loop request only the layer indices required by the DFlash assistant, reducing memory pressure without changing the assistant logic.

File: src/transformers/generation/candidate_generator.py (hunk 4 above)
Comment (sywangyi) [reply src/transformers/generation/candidate_generator.py]: ok, I will update this part

**Comment by sywangyi** [comment]
Thanks, I agree with both points.

I will removed the TP-specific slicing from `DFlashTokenCandidateGenerator` and will move it into the TP plan of `encoder.fc` 

For `output_hidden_states_layers`, the motivation is memory: DFlash only consumes `assistant_model.config.target_layer_ids`, but `output_hidden_states=True` keeps all layer outputs. This matters for MuseGlimmer long-context/multimodal generation. I agree this is a general output-capturing feature rather than MuseGlimmer-specific, so I can split it into a separate PR and keep this one focused on MuseGlimmer DFlash/TP support.

File: src/transformers/models/muse_glimmer/configuration_muse_glimmer.py (hunk 1 above)
Comment (sywangyi) [reply src/transformers/models/muse_glimmer/configuration_muse_glimmer.py:57]: yes, I did a test, without it we could not load in 2x32GB cards using 2TP for video+text case

File: src/transformers/generation/candidate_generator.py (hunk 2 above)
Comment (sywangyi) [reply src/transformers/generation/candidate_generator.py]: delete the if else

File: src/transformers/generation/candidate_generator.py [hunk 5]
```diff
@@ -1673,8 +1673,7 @@ def get_candidates(
         # Create the new inputs corresponding to only the "noise", or "diffusion window". It's the last bonus token (or "anchor") from
         # the main model, and the noise tokens
         noise_ids = torch.cat([input_ids[:, -1:], self.noise_ids_mask.to(input_ids.device)], dim=-1)
-        # The assistant needs embedding without norm thus take the lookup table and call `F.embedding`
-        noise_embeds = torch.nn.functional.embedding(noise_ids, self.main_model_input_embeddings.weight)
+        noise_embeds = self.main_model_input_embeddings(noise_ids, apply_norm=False)
```
Comment (sywangyi) [inline src/transformers/generation/candidate_generator.py]: We call the embedding module with apply_norm=False instead of using F.embedding on .weight directly because DFlash needs raw noise-token embeddings while still going through the model embedding module. This preserves TP-aware embedding dispatch/hooks; F.embedding would bypass the module and operate directly on the local weight tensor.

**Comment by sywangyi** [comment]
, allowing capture of specific layer indices is implemented in https://github.com/huggingface/transformers/pull/48081

**Review by ArthurZucker — COMMENTED** [review]
I am sorry, I fail to understand why we add code in the modeling instead of just explicitly writing stuff in the candidate generator? 

TP is not something we solve inside the modeling, we create a new TP and associate it via the TP if needed!

**Comment by sywangyi** [comment]
> I am sorry, I fail to understand why we add code in the modeling instead of just explicitly writing stuff in the candidate generator?
> 
> TP is not something we solve inside the modeling, we create a new TP and associate it via the TP if needed!

Thanks, I understand now. The `apply_norm=False` path was DFlash-specific, so it should not be exposed through MuseGlimmer modeling.

I removed that modeling API change and restored `MuseGlimmerTextNormedEmbedding.forward` to the original implementation. The raw-noise-embedding handling is now explicit in `DFlashTokenCandidateGenerator`, while the embedding module is still called normally so the TP hooks/plan remain in charge of tensor-parallel behavior.

File: src/transformers/generation/candidate_generator.py [hunk 6]
```diff
@@ -1620,6 +1621,23 @@ def __init__(
...
+
+        self.main_model_input_embeddings.embed_norm = nn.Identity()
+        try:
+            yield
+        finally:
+            self.main_model_input_embeddings.embed_norm = embed_norm
+
+    def _get_noise_embeds(self, noise_ids: torch.LongTensor) -> torch.Tensor:
+        with self._skip_embedding_norm():
+            return self.main_model_input_embeddings(noise_ids)
```
Comment (ArthurZucker) [inline src/transformers/generation/candidate_generator.py]: Mmm the context manager looks a bit like an overkill, make it a single simple function with a documentation explaining why we need this please

**Review by ArthurZucker — APPROVED** [review]

File: src/transformers/generation/candidate_generator.py (hunk 6 above)
Comment (sywangyi) [reply src/transformers/generation/candidate_generator.py]: done

**Comment by sywangyi** [comment]
@ArthurZucker I have updated the PR based on your final comment, could you merge it now?
